example/sparse_advection/parthenon_app_inputs.cpp

```diff
@@ -0,0 +1,182 @@
+  IndexRange ib = cellbounds.GetBoundsI(IndexDomain::interior);
+  IndexRange jb = cellbounds.GetBoundsJ(IndexDomain::interior);
+  IndexRange kb = cellbounds.GetBoundsK(IndexDomain::interior);
+
+  auto coords = pmb->coords;
+  auto &data = pmb->meshblock_data.Get();
+  auto pkg = pmb->packages.Get("sparse_advection_package");
+  const auto r = pkg->Param<Real>("init_size");
+  const auto size = r * r;
+
+  const auto &x0s = pkg->Param<RealArr_t>("x0");
+  const auto &y0s = pkg->Param<RealArr_t>("y0");
+
+  for (int f = 0; f < NUM_FIELDS; ++f) {
+    // allocate the sparse field on the blocks where we get non-zero values
+    bool any_nonzero = false;
+    const Real x0 = x0s[f];
+    const Real y0 = y0s[f];
+
+    for (int k = kb.s; k <= kb.e; k++) {
+      for (int j = jb.s; j <= jb.e; j++) {
+        for (int i = ib.s; i <= ib.e; i++) {
+          auto x = coords.x1v(i) - x0;
+          auto y = coords.x2v(j) - y0;
+          auto z = coords.x3v(k);
+          auto r2 = x * x + y * y + z * z;
+          if (r2 < size) {
+            any_nonzero = true;
+            break;
+          }
+        }
+        if (any_nonzero) break;
+      }
+      if (any_nonzero) break;
+    }
+
+    if (any_nonzero) {
+      pmb->AllocSparseID("sparse", f);
+      auto v = data->GetCellVariableMap().at(MakeVarLabel("sparse", f))->data;
+      pmb->par_for(
+          "SparseAdvection::ProblemGenerator", kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+          KOKKOS_LAMBDA(const int k, const int j, const int i) {
+            auto x = coords.x1v(i) - x0;
+            auto y = coords.x2v(j) - y0;
+            auto z = coords.x3v(k);
+            auto r2 = x * x + y * y + z * z;
+            v(k, j, i) = (r2 < size ? 1.0 : 0.0);
+          });
+    }
+  }
+
+  if (pkg->Param<bool>("restart_test")) {
+    bool any_nonzero = false;
+
+    for (int k = kb.s; k <= kb.e; k++) {
+      for (int j = jb.s; j <= jb.e; j++) {
+        for (int i = ib.s; i <= ib.e; i++) {
+          auto x = coords.x1v(i);
+          auto y = coords.x2v(j);
+          auto z = coords.x3v(k);
+          auto r2 = x * x + y * y + z * z;
+          if (r2 < 0.5 * size) {
+            any_nonzero = true;
+          }
+        }
+      }
+    }
+
+    if (any_nonzero) {
+      pmb->AllocSparseID("z_shape_shift", 1);
+      pmb->AllocSparseID("z_shape_shift", 3);
+      pmb->AllocSparseID("z_shape_shift", 4);
+
+      auto v = data->PackVariables(
+          std::vector<std::string>{"z_dense_A", "z_dense_B", "z_shape_shift"});
+
+      pmb->par_for(
+          "SparseAdvection::ProblemGenerator", 0, v.GetDim(4) - 1, kb.s, kb.e, jb.s, jb.e,
+          ib.s, ib.e, KOKKOS_LAMBDA(const int n, const int k, const int j, const int i) {
+            auto x = coords.x1v(i);
+            auto y = coords.x2v(j);
+            auto z = coords.x3v(k);
+            auto r2 = x * x + y * y + z * z;
+            v(n, k, j, i) = (r2 < 0.5 * size ? 1.0 : 0.0);
+          });
+    }
+  }
+}
```

Comment:
Seems like there's an awful lot of duplicate code between "restart test" and not. Can they be combined? e.g., with something like
```C++
if (restart_test) size *= 0.5;
```
for the size check and just condition the pack call on `restart_test` too? That way you only need one host-side loop and one `par_for` loop?